diff --git a/mhlo/IR/hlo_ops.cc b/mhlo/IR/hlo_ops.cc
index 8a4bef78a..39fd20b3f 100644
--- a/mhlo/IR/hlo_ops.cc
+++ b/mhlo/IR/hlo_ops.cc
@@ -5040,6 +5040,8 @@ OpFoldResult ClampOp::fold(FoldAdaptor adaptor) {
   if (!operand || !min || !max) {
     return {};
   }
+  if (!min.isSplat() || !max.isSplat())
+    return {};
   if (min.getType().getRank() == 0) {
     min = DenseElementsAttr::get(operand.getType(),
                                  min.getValues<Attribute>()[0]);
